File: 1282/solution.ts
// Run with ts-node
// $ ts-node 1282/solution.ts

const input = [3, 3, 3, 3, 3, 1, 3];

function groupThePeople(groupSizes: number[]): number[][] {
  const groupAccumulatorMap: { [key: number]: number[] } = {};
  const groups: number[][] = [];
  for (let i = 0; i < groupSizes.length; i++) {
    const groupSize = groupSizes[i];
    if (groupAccumulatorMap[groupSize] == null) {
      groupAccumulatorMap[groupSize] = [];
    }

    groupAccumulatorMap[groupSize].push(i);

    if (groupAccumulatorMap[groupSize].length === groupSize) {
      groups.push(groupAccumulatorMap[groupSize]);
      groupAccumulatorMap[groupSize] = [];
    }
  }
  return groups;
};

console.log(groupThePeople(input));

File: 1282/solution.py
import collections as c
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        result = []
        dic = {}
        
        for i, size in enumerate(groupSizes):
            
            dic.setdefault(size, []).append(i)
           
            if len(dic[size]) == size:
                result.append(dic.pop(size))
        
        return result